### PR TITLE
[fix] Fast-path skipped Context Tree IO decisions

### PR DESCRIPTION
**Pull request by Gandy2025**
## Summary

Reduces CPU work in the remaining Context Tree `io.skipped` diagnostics decision loop.

- Adds a no-ref fast path for valid tool-call candidates that cannot become recorded IO because they have no `toolFileRefs`.
- Keeps rows with refs, malformed refs, legacy `context_tree_usage`, and other non-fast-path cases on the existing schema validation and normalization path.
- Preserves existing skip-reason semantics for Claude/Codex read/write tools and shell commands.
- Adds `io_skipped_decide_fast_rows` / `io_skipped_decide_slow_rows` timing markers so dev captures can show whether the hot path is actually using the fast path.
- Adds regression coverage for no-ref Read, shell read, and shell write candidates.

## Why

After #1294 landed and deployed to dev, fresh captures from `https://dev.cloud.first-tree.ai/context` showed the SQL candidate filter helped:

- cached snapshot wall time is now roughly ~2.0-2.2s
- `io_skipped_scan` dropped to ~522-671ms
- `io_skipped_decide` still stays around ~720-740ms

The remaining cost is in the per-row decision loop. A large class of candidates has no file refs; those rows can only produce skipped diagnostics and do not need full event schema validation plus ref normalization.

## Validation

- `pnpm --filter @first-tree/server exec vitest run src/__tests__/context-tree-io.test.ts -t "fast-paths skipped diagnostics"`
- `pnpm --filter @first-tree/server exec vitest run src/__tests__/context-tree-io.test.ts`
- `pnpm --filter @first-tree/server exec vitest run src/__tests__/context-tree-snapshot-timing.test.ts src/__tests__/context-tree-snapshot.test.ts`
- `pnpm --filter @first-tree/server typecheck`
- `pnpm exec biome check packages/server/src/services/context-tree-io.ts packages/server/src/__tests__/context-tree-io.test.ts`
- `git diff --check`

### CODE REVIEW

File: packages/server/src/services/context-tree-io.ts [hunk 1]
```diff
@@ -433,40 +484,59 @@ export async function summarizeContextTreeIoSkippedEvents(
...
+
+  let fastPathRows = 0;
+  let slowPathRows = 0;
   timeSyncWithSink(
     options.timing,
     "io_skipped_decide",
     () => {
       for (const row of rows) {
+        const runtimeProvider = runtimeProviderByAgent.get(row.agentId) ?? "unknown";
+        const fastPath = skippedDecisionFastPathForNoRefs(row.kind, row.payload, runtimeProvider);
```
Comment (yuezengwu) [inline packages/server/src/services/context-tree-io.ts]: This fast path runs before `buildContextTreeIoDecision`, so it bypasses the existing `!bindingRepo -> no_org_context_tree_binding` precedence at line 322. For an org with no Context Tree binding, a no-ref `Read`/`Bash` candidate now records `no_tool_file_refs` or `unsupported_shell_command` instead of surfacing the setup/binding problem in `io.skipped`. Please gate the fast path on `binding.repo` being present, or have the fast path return `no_org_context_tree_binding`, and add a regression test using a missing/empty `contextTreeBinding`.

**Review by yuezengwu — CHANGES_REQUESTED** [review]
实现目标：这次 PR 想降低 Context Tree `io.skipped` diagnostics 决策循环的 CPU 成本，对没有 `toolFileRefs`、不可能落成 recorded IO 的 tool-call 候选走快速分类路径，并加 fast/slow row timing marker 与回归测试。

核心改动：`summarizeContextTreeIoSkippedEvents` 在 schema parse/ref normalization 之前调用 `skippedDecisionFastPathForNoRefs`，只有 rows with refs / malformed refs / legacy `context_tree_usage` 等继续走原路径。没有看到核心数据结构或数据库 schema 变更。

需要先修复一个语义回归：fast path 在缺少 org Context Tree binding 时绕过了旧路径的 `no_org_context_tree_binding` 判定，会把 setup/binding 问题误报成 no refs 或 unsupported shell command。

**Review by baixiaohang — CHANGES_REQUESTED** [review]
## Review by codex-assistant - PR #1296

**建议**: request_changes
- 理由: fast path 在缺少 Context Tree binding 时绕过了旧决策入口的 `no_org_context_tree_binding`，会把 setup/binding 问题误报成 no refs 或 unsupported shell command。

**分级**: B-low
- Path 初判: 仅改 `packages/server/**`，且不是 B-high 指定的 chat/message/session/participant/chat-membership/agent service 路径
- 语义升级: 无

**PR 摘要**
- 作者 / repo: Gandy2025 / agent-team-foundation/first-tree
- 解决问题: 降低 Context Tree `io.skipped` diagnostics 决策循环中大量 no-ref candidate 的 CPU 成本
- 设计方案: 对无 `toolFileRefs` 的 tool-call candidate 在 schema validation/ref normalization 前快速分类，并增加 fast/slow row timing marker 与回归测试
- 受影响模块: `packages/server/src/services/context-tree-io.ts`, `packages/server/src/__tests__/context-tree-io.test.ts`

**Review 意见**
❌ 1. fast path 改变了缺失 binding 时的 skip reason [R4 / packages/server/src/services/context-tree-io.ts:515]

旧 slow path 会先进入 `buildContextTreeIoDecision`，并在 `binding.repo` 缺失时立即返回 `no_org_context_tree_binding`。现在循环在传入 `binding.repo` 前先调用 `skippedDecisionFastPathForNoRefs`，因此 unbound org 里的 no-ref `Read` / `Bash` rows 会被记成 `no_tool_file_refs` 或 `unsupported_shell_command`。这会让 `io.skipped` 把 setup/binding 问题误报成 telemetry 覆盖问题。建议让 fast path 只在有 binding repo 时启用，或把 `no_org_context_tree_binding` 明确纳入 fast path，并补一个 `{ contextTreeBinding: {} }` 的 regression。

**已执行**
- 已 request_changes

**Comment by Gandy2025** [comment]
Addressed the requested no-binding regression.

Current revision:
- Fast path is now gated on an existing Context Tree binding repo, so orgs without a binding still go through the original decision path and report `no_org_context_tree_binding`.
- Added a regression test for a no-binding org with a no-ref `Read` candidate; it asserts `no_org_context_tree_binding`, `io_skipped_decide_fast_rows=0`, and `io_skipped_decide_slow_rows=1`.

Validation rerun locally:
- `pnpm --filter @first-tree/server exec vitest run src/__tests__/context-tree-io.test.ts -t "fast-paths skipped diagnostics|no-binding skipped diagnostics"`
- `pnpm --filter @first-tree/server exec vitest run src/__tests__/context-tree-io.test.ts`
- `pnpm --filter @first-tree/server exec vitest run src/__tests__/context-tree-snapshot-timing.test.ts src/__tests__/context-tree-snapshot.test.ts`
- `pnpm --filter @first-tree/server typecheck`
- `pnpm exec biome check packages/server/src/services/context-tree-io.ts packages/server/src/__tests__/context-tree-io.test.ts`
- `git diff --check`

**Review by yuezengwu — APPROVED** [review]
实现目标：这次 PR 降低 Context Tree `io.skipped` diagnostics 决策循环中大量 no-ref candidate 的 CPU 成本。

核心改动：`summarizeContextTreeIoSkippedEvents` 对有 Context Tree binding 的 no-ref tool-call candidate 先走 `skippedDecisionFastPathForNoRefs`，rows with refs / malformed refs / legacy `context_tree_usage` / no-binding cases 继续走原 schema validation 与 normalization 决策路径；同时增加 fast/slow row timing marker 和回归测试。没有看到核心数据结构或数据库 schema 变更。

复审结果：上轮 blocker 已修复。fast path 现在以 `binding.repo` 为门槛；无 binding 的 org 会回到旧路径并继续报告 `no_org_context_tree_binding`。新增 no-binding regression 覆盖了该语义。我没有发现新的 blocker，批准。

**Review by baixiaohang — APPROVED** [review]
## Review by codex-assistant - PR #1296

**建议**: approve
- 理由: 上轮 no-binding skip reason blocker 已修复，fast path 现在不会绕过缺失 Context Tree binding 的原始决策边界。

**分级**: B-low
- Path 初判: 仅改 `packages/server/**`，且不是 B-high 指定的 chat/message/session/participant/chat-membership/agent service 路径
- 语义升级: 无

**PR 摘要**
- 作者 / repo: Gandy2025 / agent-team-foundation/first-tree
- 解决问题: 降低 Context Tree `io.skipped` diagnostics 决策循环中大量 no-ref candidate 的 CPU 成本
- 设计方案: 有 binding 的 no-ref tool-call candidate 走 `skippedDecisionFastPathForNoRefs`，rows with refs / malformed refs / legacy `context_tree_usage` / no-binding cases 继续走原 schema validation 与 normalization 路径；新增 fast/slow row timing marker 和回归测试
- 受影响模块: `packages/server/src/services/context-tree-io.ts`, `packages/server/src/__tests__/context-tree-io.test.ts`

**Review 意见**
✅ 1. 上轮 blocker 已修复：`skippedDecisionFastPathForNoRefs` 现在以 `binding.repo` 为门槛，无 binding org 会回到旧路径并继续报告 `no_org_context_tree_binding`，新增 regression 覆盖 `fast_rows=0` / `slow_rows=1`。
✅ 2. 未发现新的阻塞问题；rows with refs、malformed refs、legacy `context_tree_usage` 仍保留在原 slow path 上。

**已执行**
- 已 approve
- 按 PR Review 模式未运行测试
